### PR TITLE
UNR-2888 Drop Irrelevant Outgoing RPCs

### PR DESCRIPTION
**Pull request by aleximprobable**
#### Description
Added timeout to drop outgoing RPCs. Added Offloading/Zoning authority checks, `PendingKillPending` check.

#### Release note
Added a new variable `QueuedOutgoingRPCWaitTime`
Outgoing RPCs will now be dropped if:
- More than `QueuedOutgoingRPCWaitTime ` time has passed.
- The worker is never expected to become authoritative in zoning/offloading scenario.
- The Actor is being destroyed.

#### Tests
Everything should work as before. Playtest should verify that irrelevant RPCs are dropped.

#### Documentation
See Release note

#### Primary reviewers
@jennifer-at-improbable-io, @mattyoung-improbable

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@ The format of this Changelog is based on [Keep a Changelog](https://keepachangel
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased-`x.y.z`] - 2020-xx-xx
+Added a new variable `QueuedOutgoingRPCWaitTime`. Outgoing RPCs will now be dropped if: more than `QueuedOutgoingRPCWaitTime` time has passed; the worker is never expected to become authoritative in zoning/offloading scenario; the Actor is being destroyed.
```
Comment (m-samiec) [inline CHANGELOG.md]: This should be under `features`

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 2]
```diff
@@ -624,6 +646,35 @@ void USpatialSender::TrackRPC(AActor* Actor, UFunction* Function, const RPCPaylo
 }
 #endif
 
+bool USpatialSender::WillHaveAuthorityOverActor(AActor* TargetActor, Worker_EntityId TargetEntity)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Nice

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp [hunk 3]
```diff
@@ -27,6 +27,7 @@ USpatialGDKSettings::USpatialGDKSettings(const FObjectInitializer& ObjectInitial
 	, bEnableHandover(true)
 	, MaxNetCullDistanceSquared(900000000.0f) // Set to twice the default Actor NetCullDistanceSquared (300m)
 	, QueuedIncomingRPCWaitTime(1.0f)
+	, QueuedOutgoingRPCWaitTime(10.0f)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: This seems high - any reason it's set to that?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 4]
```diff
@@ -624,6 +646,35 @@ void USpatialSender::TrackRPC(AActor* Actor, UFunction* Function, const RPCPaylo
...
+		}
+	}
+
+	if (GetDefault<USpatialGDKSettings>()->bEnableUnrealLoadBalancer)
+	{
+		if (NetDriver->VirtualWorkerTranslator != nullptr)
+		{
+			if (const SpatialGDK::AuthorityIntent* AuthorityIntentComponent = StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(TargetEntity))
+			{
+				if (AuthorityIntentComponent->VirtualWorkerId != NetDriver->VirtualWorkerTranslator->GetLocalVirtualWorkerId())
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: I wonder if this check might be too aggressive - perhaps this is set correctly, but the local worker hasn't processed the state change yet. Thoughts?

Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: If we've sent this RPC _before_ processing the state change, doesn't that mean the code would have been in a context of no authority? In which case it shouldn't have expected to be able to send the RPC?

This should catch the main case of firing RPCs on actors we've just spawned but not gained authority of yet.

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp (hunk 3 above)
Comment (mattyoung-improbable) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: Was my suggestion but yeah, probably much higher than needed. Could make it 1 second to match?

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h [hunk 5]
```diff
@@ -149,6 +149,10 @@ class SPATIALGDK_API USpatialGDKSettings : public UObject
 	UPROPERTY(EditAnywhere, config, Category = "Replication", meta = (DisplayName = "Wait Time Before Processing Received RPC With Unresolved Refs"))
 	float QueuedIncomingRPCWaitTime;
 
+	/** Seconds to wait before dropping an outgoing RPC, due to lack of authority over the Entity*/
```
Comment (mattyoung-improbable) [inline SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: Is this the only reason we would drop them? Aren't there other ways for them to time out?

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: You're right, that comment is outdated, will update.

File: CHANGELOG.md (hunk 1 above)
Comment (aleximprobable) [reply CHANGELOG.md]: Updated.

File: SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp (hunk 3 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/SpatialGDKSettings.cpp]: Changed to 1 second

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 2 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Thanks

File: SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h (hunk 5 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Public/SpatialGDKSettings.h]: Updated.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 4 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Should this be resolved?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 6]
```diff
@@ -605,24 +606,45 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
...
+	if (AActor* TargetActor = Cast<AActor>(TargetObject))
+	{
+		if (TargetActor->IsPendingKillPending())
+		{
+			bShouldDrop = true;
+		}
+
+		if (Result == ERPCResult::NoAuthority)
+		{
+			bShouldDrop = !WillHaveAuthorityOverActor(TargetActor, Params.ObjectRef.Entity);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Will this code get triggered for client -> server RPCs also? If so, the authority check should be different.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 4 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: It checks the authority of the specific RPC component, but I don't have a good use-case for when this permutation could occur. Something we should watch out for though.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 6 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: This code is used for both client and server. As I understand, the existing check `if (Result == ERPCResult::NoAuthority)` should be enough?

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Yep I see - ignore me.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 7]
```diff
@@ -605,24 +606,54 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
...
+
+	if (bShouldDrop)
+	{
+		return FRPCErrorInfo{ TargetObject, Function, NetDriver->IsServer(), ERPCQueueType::Send, ERPCResult::TimedOut, bShouldDrop };
 	}
 
 	ERPCResult Result = SendRPCInternal(TargetObject, Function, Params.Payload);
 
-	return FRPCErrorInfo{ TargetObject, Function, NetDriver->IsServer(), ERPCQueueType::Send, Result };
+	if (Result == ERPCResult::NoAuthority)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Move bShouldDrop declaration to here

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 8]
```diff
@@ -605,24 +606,54 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
 {
 	SCOPE_CYCLE_COUNTER(STAT_SpatialSenderSendRPC);
 
+	bool bShouldDrop = false;
+
+	const float TimeDiff = (FDateTime::Now() - Params.Timestamp).GetTotalSeconds();
+	if (GetDefault<USpatialGDKSettings>()->QueuedOutgoingRPCWaitTime < TimeDiff)
+	{
+		bShouldDrop = true;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Don't bother with should drop, just return immediately

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 9]
```diff
@@ -605,24 +606,54 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
...
 	{
-		return FRPCErrorInfo{ TargetObject, nullptr, NetDriver->IsServer(), ERPCQueueType::Send, ERPCResult::MissingFunctionInfo };
+		return FRPCErrorInfo{ TargetObject, nullptr, NetDriver->IsServer(), ERPCQueueType::Send, ERPCResult::MissingFunctionInfo, bShouldDrop };
+	}
+
+	if (AActor* TargetActor = Cast<AActor>(TargetObject))
+	{
+		if (TargetActor->IsPendingKillPending())
+		{
+			bShouldDrop = true;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Just return immediately

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 10]
```diff
@@ -605,24 +606,54 @@ FRPCErrorInfo USpatialSender::SendRPC(const FPendingRPCParams& Params)
...
+			return FRPCErrorInfo{ TargetObject, Function, NetDriver->IsServer(), ERPCQueueType::Send, ERPCResult::ActorPendingKill, bShouldDrop };
+		}
+	}
+
+	if (bShouldDrop)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Then we can remove this

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 8 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: In that case logs will not contain the information about the TargetActor, Function etc. Is that acceptable?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp (hunk 9 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]: Can do, I had this variable here for readability. Will pass `true` as one of the arguments on the next line instead.

**Pull request merged**